### ISSUE TITLE
Lazier LazyList?

### ISSUE THREAD

**Issue by julienrf** [scala/collection-strawman#367]
@sjrd pointed mo to https://github.com/scala/bug/issues/10696

In our current `LazyList` implementation, you have to say whether a `LazyList` instance is going to be empty or non empty at the time you build it, making it impossible to implement an unblocking `linesStream` operation. We should consider delaying the moment where we say whether the `LazyList` is empty or not to iteration time instead of construction time.

**Comment by szeiger** [comment]
It's getting too late in the 2.13 release cycle for this but it would be nice to reconsider it for 2.14.

**Comment by NthPortal** [comment]
It doesn't seem to me like it ought to be that difficult to change. Currently, the two concrete subclasses of `LazyList` (like `List`) are `Empty` and `Cons`. It doesn't seem that far-fetched to me to add an `Unknown` subclass of some sort which doesn't know what it is yet, but will resolve to an `Empty` or `Cons` (probably via wrapping) once it does know whether it has elements.

**Comment by NthPortal** [comment]
> making it impossible to implement an unblocking `linesStream` operation

I'm not sure that deferring computation of whether or not the `LazyList` is empty to iteration time will enable non-blocking IO - at the end of the day, the lines are either there when you try to compute them or not. Please correct me if I'm misunderstanding

**Comment by sjrd** [comment]
Well for one, it can't be a sealed trait with two concrete subclasses for empty and non-empty, since that immediately requires to synchronously decide which subclass to return. 

Instead, the definition of `LazyList` should look like the following:
```scala
object LazyList {
  private sealed abstract class State[+A]
  private case object Empty extends State[Nothing]
  private final class NonEmpty[+A](byNameHead: => A, val tail: LazyList[A]) extends State[A] {
    lazy val head: A = byNameHead
  }
}

final class `LazyList[+A] private (byNameState: => State[A]) {
  private lazy val state: State[A] = byNameState
}
```
or something else with equivalent expressivity.

The good news is that the public API should be mostly unchanged, except that of course we need at least one public "constructor" that exposes the lazy nature of empty/non-empty.

**Comment by NthPortal** [comment]
> The good news is that the public API should be mostly unchanged, except that of course we need at least one public "constructor" that exposes the lazy nature of empty/non-empty.

I think `from[A](IterableOnce[A]): LazyList[A]` does that pretty trivially. Just wait to call `hasNext` on the iterator.

**Comment by sjrd** [comment]
@NthPortal it won't be *non-blocking* I/O in the sense that it's all using asynchronous I/O without locks, of course. But at least you can actually make the start of I/O operation lazy, like the rest of it.

But that's not even the point. This problem is not related to I/O. It's much larger than that! If you call `xs.filter(f)` where `xs` is a `LazyList`, you would expect the call to be lazy and immediately return, wouldn't you? Problem is, it's *not*, because it needs to decide right away whether to return an empty or non-empty one. Which, in the case of `filter`, requires an unbounded number of operations.

**Comment by NthPortal** [comment]
Potentially we could just tweak `Deferrer` to be a subclass of `LazyList`?

**Comment by adriaanm** [comment]
Let’s fix this. This is our shot at fixing Stream

**Comment by mdedetrich** [comment]
> If this really can't be fixed by 2.13.0 it seems like it'd be more responsible to remove LazyList from the collections than ship something known to be broken.

Strongly agreed, I would rather fix this properly now. The current collections already have a bad enough rep for our broken `Stream` type, we really shouldn't repeat history in this regard
